### PR TITLE
Compare: move `get_similarities_at_index` into `compare_parallel` function to help with memory

### PR DESCRIPTION
**Pull request by olgabot**
When running `compare` on large (~30,000 signature) datasets, even on beefy machines, the memory can lock up (ref: https://github.com/dib-lab/sourmash/issues/1486). @pranathivemuri added this fix, which moves the function `get_similarities_at_index` into the inside of `compare_parallel`, which reduces the amount of serialization passed between threads. I'm not 100% sure on how this works, @pranathivemuri is the expert here!

### CODE REVIEW

**Comment by ctb** [comment]
yes, this sounds like something I had to fix when doing releases - you need
to install a package.

I think it's setuptools_scm_git_archive, installable with pip. I'll go find
the issue and link to it later when I get a chance!

On Sat, May 08, 2021 at 10:32:20AM -0700, Olga Botvinnik wrote:
> For some reason the version appears as `0.0.0` -- any ideas why?
> 
> ```
> (nf-core--kmermaid-1.1.0dev)
>  Sat  8 May - 10:31 ??? ~/code/sourmash ??? ??? origin ??? olgabot/pranthi-fxn-in-fxn ??? ???
>  ***@***.*** ??? sourmash info
> 
> == This is sourmash version 0.0.0. ==
> == Please cite Brown and Irber (2016), doi:10.21105/joss.00027. ==
> 
> sourmash version 0.0.0
> - loaded from path: /data_sm/home/olga_ibm/miniconda3/envs/nf-core--kmermaid-1.1.0dev/lib/python3.7/site-packages/sourmash-0.0.0-py3.7-linux-ppc64le.egg/sourmash/cli
> ```
> 
> -- 
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/dib-lab/sourmash/pull/1509#issuecomment-835439531
-- 
C. Titus Brown, ***@***.***

**Comment by ctb** [comment]
the version stuff can be fixed with `pip install setuptools_scm`, I believe. Here's the link to when it first popped up in release 4.0.0rc1: https://github.com/dib-lab/sourmash/issues/1337#issuecomment-781369145

File: src/sourmash/compare.py [hunk 1]
```diff
@@ -159,15 +127,37 @@ def compare_parallel(siglist, ignore_abundance, downsample, n_jobs):
...
-    # siglist, ignore_abundance, downsample, for computing all the signatures
-    # The only changing parameter that will be mapped from the pool is the index
-    func = partial(
-        get_similarities_at_index,
-        siglist=siglist,
-        ignore_abundance=ignore_abundance,
-        downsample=downsample)
-    notify("Created similarity func")
+    # To avoid sharing siglist via pickle declaring a function inside function
+
```
Comment (pranathivemuri) [inline src/sourmash/compare.py]:
```suggestion
global get_similarities_at_index
```

File: src/sourmash/compare.py [hunk 2]
```diff
@@ -159,15 +127,37 @@ def compare_parallel(siglist, ignore_abundance, downsample, n_jobs):
...
-    # siglist, ignore_abundance, downsample, for computing all the signatures
-    # The only changing parameter that will be mapped from the pool is the index
-    func = partial(
-        get_similarities_at_index,
-        siglist=siglist,
-        ignore_abundance=ignore_abundance,
-        downsample=downsample)
-    notify("Created similarity func")
+    # To avoid sharing siglist via pickle declaring a function inside function
+global get_similarities_at_index
```
Comment (pranathivemuri) [inline src/sourmash/compare.py]:
```suggestion
    global get_similarities_at_index
```

**Comment by pranathivemuri** [comment]
@olgabot could you please try now? By using nested function and the shared variables in the outer function I am trying to avoid serialization of `siglist` that seems to take up a lot of memory as per your issue #1486

**Comment by pranathivemuri** [comment]
@ctb @olgabot should we rebase and get a review?

**Comment by ctb** [comment]
> @ctb @olgabot should we rebase and get a review?

yes, please!

**Comment by ctb** [comment]
I might pick this up now that I grok some of the code more b/c of #2265. See also expert commentary [here](https://github.com/numpy/numpy/issues/22213#issuecomment-1238263592).

**Pull request closed without merging**